### PR TITLE
Detail how to debug CI failures as a contributor

### PR DESCRIPTION
**Pull request by mariantalla**
This PR adds more detail on what contributors can do to debug CI failures and flakes on their contributions.

Tasks:
- [x] What to do if failure is not related to the changes of the PR
  - [x] Triaging failure
  - [x] Opening an issue
  - [x] Notifying SIG

Is it worth, as part of this PR:
- Including an overview of testgrid? Other tooling?

Related:
- https://github.com/kubernetes/community/issues/1537
- https://github.com/kubernetes/sig-release/issues/428

### CODE REVIEW

**Comment by BenTheElder** [comment]
thanks @mariantalla -- I'm pretty swamped at the moment, mind pinging me sometime next week to revisit? I would love to help add some details :sweat_smile:

File: contributors/devel/testing.md [hunk 1]
```diff
@@ -225,3 +225,40 @@ version and the watch cache test is skipped.
...
+
+### Troubleshooting a failure
+Click on `Details` and look at the [`gubernator`](gubernator.k8s.io/) output for the test.
+
+#### Troubleshooting failures/flakes that are not caused by your change
+- Check if an issue has already been opened for that test failure/flake
+  - If yes, comment on it and link your PR, the failed run that affected you and any other information you think might be relevant
+  - If no, open a new issue and notify the appropriate SIG (see: SIG test escalation)
+
+#### SIG test escalation
```
Comment (neolit123) [inline contributors/devel/testing.md]: i would title this:
> Escalating failures to a SIG

File: contributors/devel/testing.md [hunk 2]
```diff
@@ -225,3 +225,40 @@ version and the watch cache test is skipped.
...
+- Figure out corresponding sig from test name/description
+- Mention the sig's github handle on the issue, optionally cc the SIG's chair(s) (locate them under kubernetes/community/sig-<name>)
```
Comment (neolit123) [inline contributors/devel/testing.md]: some capitalizations:

`sig` -> `SIG`
`github` -> `GitHub`
`cc` -> `CC`

File: contributors/devel/testing.md [hunk 3]
```diff
@@ -225,3 +225,40 @@ version and the watch cache test is skipped.
 ## End-to-End tests
 
 Please refer to [End-to-End Testing in Kubernetes](e2e-tests.md).
+
+## Running your contribution in the Kubernetes CI
+Once you open a PR, prow will run pre-submit tests in CI.
```
Comment (neolit123) [inline contributors/devel/testing.md]: possibly mention `/ok-to-test` and `/retest` and/or link to some of the PR / test related bot commands?

File: contributors/devel/testing.md [hunk 4]
```diff
@@ -225,3 +225,40 @@ version and the watch cache test is skipped.
...
+
+### Troubleshooting a failure
+Click on `Details` and look at the [`gubernator`](gubernator.k8s.io/) output for the test.
+
+#### Troubleshooting failures/flakes that are not caused by your change
```
Comment (neolit123) [inline contributors/devel/testing.md]: possibly unify the two sections?

we can explain that if the gubernator output seems unrelated to the change they might try to first call `/retest` them self. but if a test continues to fail, to contact a reviewer/maintainer.

usually what happens is that contributors first ping in the PR and if nobody responds they go asking on slack.

File: contributors/devel/testing.md (hunk 1 above)
Comment (mariantalla) [reply contributors/devel/testing.md]: Thanks, changed it!

File: contributors/devel/testing.md (hunk 2 above)
Comment (mariantalla) [reply contributors/devel/testing.md]: Thanks, capitalized sig and github except where (I thought) it made sense to keep lower-case (e.g. in urls)

I changed cc to code format (`cc`) rather than capitalizing with the logic that it's sort of a command within a GitHub conversation... no strong opinions though. What do you think?

Comment (neolit123) [reply contributors/devel/testing.md]: sounds good :+1:

File: contributors/devel/testing.md (hunk 3 above)
Comment (mariantalla) [reply contributors/devel/testing.md]: Added, thanks!

File: contributors/devel/testing.md (hunk 4 above)
Comment (mariantalla) [reply contributors/devel/testing.md]: Good point, we unified them under a single title and structured them a bit more.

**Comment by mariantalla** [comment]
/cc @spiffxp @fejta @stevekuznetsov @timothysc @tpepper 

(`sig-testing-leads` and https://github.com/kubernetes/community/issues/1537 creator)

File: contributors/devel/sig-testing/testing.md [hunk 5]
```diff
@@ -197,3 +197,61 @@ Please refer to [Integration Testing in Kubernetes](integration-tests.md).
...
+
+## Running your contribution in the Kubernetes CI
+Once you open a PR, [`prow`][prow-doc] runs pre-submit tests in CI.
+
+If you are not a [Kubernetes org member][member], another org member will need to run [`/ok-to-test`][ok-to-test] on your PR.
+
+Find out more about [other commands][prow-cmds] you can use to interact with prow through GitHub comments.
+
+### Troubleshooting a failure
+Click on `Details` and look at the [`gubernator`](gubernator.k8s.io/) output for the test.
```
Comment (spiffxp) [inline contributors/devel/sig-testing/testing.md]: I would reword this as we are looking to move away from gubernator.  It also doesn't make much sense to a new contributor.  What am I going here to look for? Artifacts produced by the test and cluster under test, such as:
- test results
- more metadata about the tests (what versions of things were used? how long did they take?)
- output from tests that have failed
- build log showing the full test run
- logs from the cluster under test (k8s components such as kublet and apiserver, possibly other logs such as etcd and kernel)
- junit xml files
- coverage files
- etc.

File: contributors/devel/sig-testing/testing.md [hunk 6]
```diff
@@ -197,3 +197,61 @@ Please refer to [Integration Testing in Kubernetes](integration-tests.md).
...
+
+If the failure seems unrelated to the change you're submitting:
+- Is it a flake?
+  - Check if an issue has already been opened for that flake
+    - If not, open a new one (like [this example][new-issue-example]) and [label it `kind/flake`][kind/flake]
+  - Run [`/retest`][retest] on your PR to re-trigger the tests
+
+- Is it a failure that shouldn't be happening (in other words, is the test now wrong)
+  - Get in touch with the SIG
```
Comment (spiffxp) [inline contributors/devel/sig-testing/testing.md]: Which SIG?

File: contributors/devel/sig-testing/testing.md [hunk 7]
```diff
@@ -197,3 +197,61 @@ Please refer to [Integration Testing in Kubernetes](integration-tests.md).
...
+Find out more about [other commands][prow-cmds] you can use to interact with prow through GitHub comments.
+
+### Troubleshooting a failure
+Click on `Details` and look at the [`gubernator`](gubernator.k8s.io/) output for the test.
+
+If the failure seems unrelated to the change you're submitting:
+- Is it a flake?
+  - Check if an issue has already been opened for that flake
+    - If not, open a new one (like [this example][new-issue-example]) and [label it `kind/flake`][kind/flake]
+  - Run [`/retest`][retest] on your PR to re-trigger the tests
```
Comment (spiffxp) [inline contributors/devel/sig-testing/testing.md]: If it is a flake, and there is an issue for it, is there anything else I can do to aid with troubleshooting beyond running `/retest` on my PR?

eg: I personally use N=1 to say fine it's a flake, N=2 to spam retest, N=3 to say OK what can I do to help identify this flake and make it go away.  Sometimes that's looking at the actual failure and seeing if it's the same failure each time it happens.  Maybe it's looking back to see when this flake started to happen, and possibly identify the window of commits that should be explored further.  etc.

File: contributors/devel/sig-testing/testing.md [hunk 8]
```diff
@@ -197,3 +197,61 @@ Please refer to [Integration Testing in Kubernetes](integration-tests.md).
...
+
+- Is it a failure that shouldn't be happening (in other words, is the test now wrong)
+  - Get in touch with the SIG
+    - preferably as a comment on your PR, by tagging the [team][k-teams] (for example a [reviewers team for the SIG][k-teams-review])
```
Comment (spiffxp) [inline contributors/devel/sig-testing/testing.md]: anecdotally, does the use of these teams work as described here?

**Comment by fejta** [comment]
/uncc
Looks great, looks like lots of activity. please add me back if I'm needed here!

File: contributors/devel/sig-testing/testing.md [hunk 9]
```diff
@@ -197,3 +197,61 @@ Please refer to [Integration Testing in Kubernetes](integration-tests.md).
...
+    - if you don't get a response in 24h, engage with the SIG on their channel on the Kubernetes slack and/or attend one of the [SIG meetings][sig-meetings] to ask for input.
+
+[prow-doc]: https://kubernetes.io/blog/2018/08/29/the-machines-can-do-the-work-a-story-of-kubernetes-testing-ci-and-automating-the-contributor-experience/#enter-prow
```
[the remark this answers is not in the document]
Comment (mariantalla) [reply contributors/devel/sig-testing/testing.md]: Good point, changed it!

File: contributors/devel/sig-testing/testing.md (hunk 5 above)
Comment (mariantalla) [reply contributors/devel/sig-testing/testing.md]: Thanks, removed the reference to gubernator and added specifics on what specifically they can find under `Details` that is useful to debugging. I haven't gone into a lot of detail as to _how_ to read through logs etc (I worry it might get too tiresome for folks to read & follow), any opinions on this?

File: contributors/devel/sig-testing/testing.md (hunk 6 above)
Comment (mariantalla) [reply contributors/devel/sig-testing/testing.md]: Good point, I _think_ the "sig-on-PR-label" is probably the best bet - is that right?

Changed the text to say that for now.

File: contributors/devel/sig-testing/testing.md (hunk 7 above)
Comment (mariantalla) [reply contributors/devel/sig-testing/testing.md]: Thanks, good point! I added a small section with pointers to what contributors can do - similar to working through artifacts I was cautious of being too prescriptive with a course of action that changes a lot depending on the context of the specific flake occurrence. Happy to add more though if you feel this is the right place to put more information.

File: contributors/devel/sig-testing/testing.md (hunk 8 above)
Comment (mariantalla) [reply contributors/devel/sig-testing/testing.md]: Hm, interesting. @hoegaarden and I have had mixed experiences so far, i.e. it's not much overhead & it hasn't hurt to tag reviewers, it sometimes helped (but sometimes made no significant difference). Not sure how representative this is though; would you recommend a different suggestion here? 🤔

Comment (spiffxp) [reply contributors/devel/sig-testing/testing.md]: no particular suggestion, I wish we had ways of measuring efficacy here

**Comment by spiffxp** [comment]
/approve
/lgtm
Thanks for making this better
